<a href="https://colab.research.google.com/github/2303A51528/AI_LAB_1528/blob/main/7.5_assmnt_1528.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### TASK-1 The Mutable Default Argument Bug

When a mutable object (like a list or dictionary) is used as a default argument in a Python function, it's created only once when the function is defined, not each time the function is called. This means that all calls to the function that don't explicitly provide a value for that argument will share the *same* mutable object, leading to unexpected behavior.

In [1]:
# Original buggy function
def add_item_buggy(item, items=[]):
    items.append(item)
    return items

print("Buggy function output:")
print(f"Call 1: {add_item_buggy(1)}")
print(f"Call 2: {add_item_buggy(2)}")
print(f"Call 3: {add_item_buggy(3)}")


Buggy function output:
Call 1: [1]
Call 2: [1, 2]
Call 3: [1, 2, 3]


As you can see from the output of the buggy function, the `items` list is not re-initialized for each call, but rather modified repeatedly. The expected behavior is that each call to `add_item` without providing the `items` argument should start with a new, empty list.

### The Fix

The standard way to fix this is to use `None` as the default argument and then initialize an empty list inside the function if `items` is `None`. This ensures that a new list is created every time the function is called without an explicit list being passed.

In [4]:
# Corrected function
def add_item_fixed(item, items=None):
    if items is None:
        items = []
    items.append(item)
    return items

print("\nCorrected function output:")
print(f"Call 1: {add_item_fixed(1)}")
print(f"Call 2: {add_item_fixed(2)}")
print(f"Call 3: {add_item_fixed(3)}")

# Demonstrate with an explicit list
my_list = ['a', 'b']
print(f"\nCall with explicit list: {add_item_fixed('c', my_list)}")
print(f"Original list after call: {my_list}")



Corrected function output:
Call 1: [1]
Call 2: [2]
Call 3: [3]

Call with explicit list: ['a', 'b', 'c']
Original list after call: ['a', 'b', 'c']


#Observation

(Mutable Default Argument Bug) has been successfully addressed! We demonstrated how using a mutable object as a default argument leads to unexpected shared state, and we fixed it by using None as the default and initializing an empty list inside the function.

###TASK-2 The Floating-Point Precision Error

Floating-point numbers are represented in computers using a finite number of binary digits, which can lead to small inaccuracies. This means that some decimal numbers cannot be represented exactly in binary, causing what's known as a floating-point precision error. When these slightly imprecise numbers are used in calculations, the results can also be slightly off, making direct equality comparisons (`==`) unreliable.

In [5]:
# Original buggy function
def check_sum_buggy():
    return (0.1 + 0.2) == 0.3

print("Buggy function output:")
print(f"Is 0.1 + 0.2 equal to 0.3? {check_sum_buggy()}")
print(f"The actual value of 0.1 + 0.2 is: {0.1 + 0.2}")


Buggy function output:
Is 0.1 + 0.2 equal to 0.3? False
The actual value of 0.1 + 0.2 is: 0.30000000000000004


As you can see, the direct comparison returns `False`, even though mathematically `0.1 + 0.2` should be `0.3`. This is because `0.1 + 0.2` actually results in a number like `0.30000000000000004`, which is not *exactly* `0.3`.

### The Fix

To correctly compare floating-point numbers, we should check if their absolute difference is less than a small tolerance value (often called epsilon). This accounts for the minor precision errors.

In [6]:
# Corrected function using a tolerance
def check_sum_fixed(tolerance=1e-9):
    return abs((0.1 + 0.2) - 0.3) < tolerance

print("\nCorrected function output:")
print(f"Is 0.1 + 0.2 approximately equal to 0.3? {check_sum_fixed()}")

# Using Python's math.isclose for a more robust check (Python 3.5+)
import math
def check_sum_isclose():
    return math.isclose(0.1 + 0.2, 0.3)

print(f"Using math.isclose: {check_sum_isclose()}")



Corrected function output:
Is 0.1 + 0.2 approximately equal to 0.3? True
Using math.isclose: True


#observation


Task 2 (Floating-Point Precision Error) has been successfully addressed! We demonstrated how direct equality comparisons with floating-point numbers can be unreliable due to precision issues, and provided solutions using a custom tolerance and Python's math.isclose() function.

### TASK-3 The Recursion Error (Missing Base Case)

A recursive function calls itself repeatedly until a certain condition (the 'base case') is met. Without a base case, the function will call itself indefinitely, eventually leading to a `RecursionError` as the call stack overflows.

In [7]:
# Original buggy function (will cause a RecursionError)
def countdown_buggy(n):
    print(n)
    return countdown_buggy(n-1)

print("Buggy function output (will eventually raise RecursionError):")
# To prevent an actual crash in the notebook, I will wrap this in a try-except block.
try:
    countdown_buggy(5)
except RecursionError as e:
    print(f"Caught an error: {e}")


Buggy function output (will eventually raise RecursionError):
5
4
3
2
1
0
-1
-2
-3
-4
-5
-6
-7
-8
-9
-10
-11
-12
-13
-14
-15
-16
-17
-18
-19
-20
-21
-22
-23
-24
-25
-26
-27
-28
-29
-30
-31
-32
-33
-34
-35
-36
-37
-38
-39
-40
-41
-42
-43
-44
-45
-46
-47
-48
-49
-50
-51
-52
-53
-54
-55
-56
-57
-58
-59
-60
-61
-62
-63
-64
-65
-66
-67
-68
-69
-70
-71
-72
-73
-74
-75
-76
-77
-78
-79
-80
-81
-82
-83
-84
-85
-86
-87
-88
-89
-90
-91
-92
-93
-94
-95
-96
-97
-98
-99
-100
-101
-102
-103
-104
-105
-106
-107
-108
-109
-110
-111
-112
-113
-114
-115
-116
-117
-118
-119
-120
-121
-122
-123
-124
-125
-126
-127
-128
-129
-130
-131
-132
-133
-134
-135
-136
-137
-138
-139
-140
-141
-142
-143
-144
-145
-146
-147
-148
-149
-150
-151
-152
-153
-154
-155
-156
-157
-158
-159
-160
-161
-162
-163
-164
-165
-166
-167
-168
-169
-170
-171
-172
-173
-174
-175
-176
-177
-178
-179
-180
-181
-182
-183
-184
-185
-186
-187
-188
-189
-190
-191
-192
-193
-194
-195
-196
-197
-198
-199
-200
-201
-202
-203
-204
-205
-206
-207

As you can see, the buggy function attempts to call itself indefinitely. Python has a default recursion limit (usually 1000 calls) to prevent infinite loops from consuming all system resources, which is why a `RecursionError` is raised.

### The Fix

The fix is to add a base case that stops the recursion. For a countdown, the natural base case is when `n` reaches 0 or less. At this point, the function should simply return without making another recursive call.

In [9]:
# Corrected function with a base case
def countdown_fixed(n):
    if n <= 0:  # Base case: stop when n is 0 or less
        print("Countdown finished!")
        return
    print(n)
    return countdown_fixed(n-1)

print("\nCorrected function output:")
countdown_fixed(5)



Corrected function output:
5
4
3
2
1
Countdown finished!


## observation-
(Recursion Error – Missing Base Case) has been successfully addressed! We demonstrated how the absence of a base case leads to infinite recursion and a RecursionError. The fix involved adding a clear stopping condition (when n <= 0) to prevent the function from calling itself indefinitely.

```
# This is formatted as code
```


```



### TASK-4 The Dictionary Key Error

A `KeyError` occurs when you try to access a key in a dictionary that does not exist. Unlike lists where an `IndexError` is raised for out-of-bounds indices, dictionaries raise a `KeyError`.

In [10]:
# Original buggy function (will cause a KeyError)
def get_value_buggy():
    data = {"a": 1, "b": 2}
    return data["c"]

print("Buggy function output (will eventually raise KeyError):")
# To prevent an actual crash in the notebook, I will wrap this in a try-except block.
try:
    print(f"Value for key 'c': {get_value_buggy()}")
except KeyError as e:
    print(f"Caught an error: {e}")


Buggy function output (will eventually raise KeyError):
Caught an error: 'c'


As you can see, attempting to access `data['c']` results in a `KeyError` because 'c' is not a key in the `data` dictionary.

### The Fix

There are two common ways to handle missing dictionary keys gracefully:

1.  **Using `dict.get(key, default_value)`:** This method returns the value for the specified `key` if it exists. If the `key` is not found, it returns `None` by default, or a `default_value` if provided.
2.  **Using a `try-except` block:** You can wrap the dictionary access in a `try` block and catch the `KeyError` if it occurs, allowing you to handle the error condition explicitly.

In [12]:
# Corrected function using .get() method
def get_value_get(key, default_value=None):
    data = {"a": 1, "b": 2}
    return data.get(key, default_value)

print("\nCorrected function output (using .get()):")
print(f"Value for existing key 'a': {get_value_get('a')}")
print(f"Value for non-existing key 'c' (default None): {get_value_get('c')}")
print(f"Value for non-existing key 'd' (default 0): {get_value_get('d', 0)}")

# Corrected function using try-except block
def get_value_try_except(key):
    data = {"a": 1, "b": 2}
    try:
        return data[key]
    except KeyError:
        return f"Key '{key}' not found!"

print("\nCorrected function output (using try-except):")
print(f"Value for existing key 'a': {get_value_try_except('a')}")
print(f"Value for non-existing key 'c': {get_value_try_except('c')}")



Corrected function output (using .get()):
Value for existing key 'a': 1
Value for non-existing key 'c' (default None): None
Value for non-existing key 'd' (default 0): 0

Corrected function output (using try-except):
Value for existing key 'a': 1
Value for non-existing key 'c': Key 'c' not found!


#  #observation
  
  
  (Dictionary Key Error) has been successfully completed! We identified the KeyError that occurs when accessing a non-existent dictionary key and demonstrated two robust solutions:

Using dict.get(): This method safely retrieves a value, returning None or a specified default if the key is not found.
Using a try-except block: This allows for explicit handling of the KeyError to provide custom feedback or alternative logic.

### TASK-5 The Infinite Loop (Wrong Condition)

An infinite loop occurs when the condition that controls a loop (like a `while` loop) never evaluates to `False`. This causes the loop to execute indefinitely, consuming system resources and potentially crashing the program or making it unresponsive. A common reason for this is forgetting to update the loop's control variable within the loop body.

In [15]:
# Original buggy function (will cause an infinite loop)
def loop_example_buggy():
    i = 0
    while i < 5:
        print(i)

print("Buggy function output (will run infinitely if not stopped manually):")
# To prevent an actual crash in the notebook, I will only show a partial run or wrap it if possible.
# For demonstration, we will let it print a few times and then expect it to run indefinitely.

# NOTE: Running this directly without modification will cause an infinite loop.
# For safety and to prevent the notebook from crashing, I will just display the problematic code
# and then proceed to the corrected version.

# print(loop_example_buggy()) # Uncomment to see the infinite loop (run at your own risk!)
print("The buggy function 'loop_example_buggy()' would print '0' repeatedly forever because 'i' is never incremented.")
print("Expected behavior: The loop should run 5 times, printing 0, 1, 2, 3, 4.")


Buggy function output (will run infinitely if not stopped manually):
The buggy function 'loop_example_buggy()' would print '0' repeatedly forever because 'i' is never incremented.
Expected behavior: The loop should run 5 times, printing 0, 1, 2, 3, 4.


As explained, the `loop_example_buggy()` function would print `0` indefinitely because the variable `i` is initialized to `0` and never changes. Therefore, `i < 5` always remains `True`.

### The Fix

To fix an infinite loop caused by a missing update, we need to ensure that the loop's control variable is modified within the loop body in a way that eventually makes the loop condition `False`. In this case, incrementing `i` in each iteration will resolve the issue.

In [16]:
# Corrected function
def loop_example_fixed():
    i = 0
    while i < 5:
        print(i)
        i += 1  # Increment i to ensure the loop terminates

print("\nCorrected function output:")
loop_example_fixed()



Corrected function output:
0
1
2
3
4


#observation

The code has executed successfully, and the outputs clearly demonstrate the infinite loop issue and its resolution:

Buggy Function Output: As I predicted, running loop_example_buggy() would have printed 0 indefinitely because the variable i was never incremented, thus i < 5 always remained True. For safety, I only displayed an explanation rather than letting it run.

Corrected Function Output: In contrast, loop_example_fixed() successfully printed 0, 1, 2, 3, 4 and then terminated. This confirms that incrementing i within the loop allowed the condition i < 5 to eventually become False, thereby preventing the infinite loop.

### TASK-6 The Unpacking Error (Wrong Number of Variables)

An unpacking error occurs when you try to assign values from an iterable (like a tuple, list, or string) to a sequence of variables, but the number of variables does not match the number of items in the iterable. Python will raise a `ValueError` in such cases.

In [19]:
# Original buggy code (will cause a ValueError)
def unpack_buggy():
    # Trying to unpack 3 items into 2 variables
    a, b = (1, 2, 3)
    print(f"a: {a}, b: {b}")

print("Buggy code output (will eventually raise ValueError):")
# To prevent an actual crash, I will wrap this in a try-except block.
try:
    unpack_buggy()
except ValueError as e:
    print(f"Caught an error: {e}")


Buggy code output (will eventually raise ValueError):
Caught an error: too many values to unpack (expected 2)


As you can see, the buggy code attempts to unpack a tuple of three elements `(1, 2, 3)` into two variables `a` and `b`. Python correctly raises a `ValueError` because there's a mismatch in the number of values.

### The Fix

To fix an unpacking error, ensure that the number of variables on the left-hand side matches the number of items in the iterable on the right-hand side. If you want to ignore certain values, you can use the underscore `_` as a placeholder variable. For multiple extra values, you can use `*` with `_` or another variable name.

In [20]:
# Corrected unpacking - matching the number of variables
def unpack_fixed_match():
    a, b, c = (1, 2, 3)
    print(f"Correct unpacking (match variables to items): a: {a}, b: {b}, c: {c}")

# Corrected unpacking - using _ for an unwanted value
def unpack_fixed_ignore_one():
    a, b, _ = (1, 2, 3) # Ignore the third value
    print(f"Correct unpacking (ignore one value with _): a: {a}, b: {b}")

# Corrected unpacking - using * for multiple unwanted values (Python 3+)
def unpack_fixed_ignore_many():
    a, *_, c = (1, 2, 3, 4, 5) # Assign first and last, ignore middle values
    print(f"Correct unpacking (ignore many values with *_): a: {a}, c: {c}")

print("\nCorrected function output:")
unpack_fixed_match()
unpack_fixed_ignore_one()
unpack_fixed_ignore_many()



Corrected function output:
Correct unpacking (match variables to items): a: 1, b: 2, c: 3
Correct unpacking (ignore one value with _): a: 1, b: 2
Correct unpacking (ignore many values with *_): a: 1, c: 5


# observatin

 (Unpacking Error – Wrong Number of Variables) has been successfully addressed! We demonstrated how a ValueError occurs when the number of variables for unpacking doesn't match the iterable's length. The solutions involved:

Matching the number of variables to the items.
Using _ to ignore specific unwanted values.
Using *_ to ignore multiple unwanted values (for flexible unpacking).

### TASK-7 The Mixed Indentation Error (Tabs vs Spaces)

Python enforces strict rules regarding indentation to define code blocks. Mixing tabs and spaces for indentation within the same file or even within the same block can lead to an `IndentationError` or subtle, hard-to-debug logical errors, especially across different text editors or environments. The Python style guide (PEP 8) recommends using 4 spaces per indentation level.

As explained, the buggy code above intentionally uses mixed indentation (spaces for `x = 5` and a tab for `y = 10`), which Python will flag as an `IndentationError` or `SyntaxError` depending on the exact interpreter and file settings. Python expects consistent indentation within a block.

### The Fix

The fix is to ensure that all indentation within a code block (and ideally, throughout the entire file) uses the same character type (either spaces or tabs) and the same level (e.g., 4 spaces per level). Most modern Python development environments automatically convert tabs to spaces or enforce consistent indentation.

In [24]:
# Corrected function with consistent indentation (using 4 spaces)
def func_fixed():
    x = 5
    y = 10
    return x + y

print("\nCorrected function output:")
print(f"Result: {func_fixed()}")



Corrected function output:
Result: 15


#observation

 (Mixed Indentation Error – Tabs vs Spaces) has been successfully addressed! We demonstrated how the inconsistent use of tabs and spaces for indentation leads to a TabError or IndentationError in Python. The fix involved ensuring consistent indentation, typically using 4 spaces per level, which allowed the code to execute correctly.

### TASK-8 The Import Error (Wrong Module Usage)

An `ImportError` (or sometimes `ModuleNotFoundError`) occurs when Python cannot find or load the module you are trying to import. This can happen for several reasons, such as:

*   **Typo in the module name:** The most common cause.
*   **Module not installed:** If it's a third-party library, it needs to be installed (e.g., via `pip`).
*   **Incorrect path:** Python's interpreter can't find the module file in its search paths.

In this task, we'll focus on a common typo.

In [25]:
# Original buggy code (will cause an ImportError)
def calculate_sqrt_buggy():
    import maths # Incorrect module name
    return maths.sqrt(16)

print("Buggy code output (will eventually raise ImportError):")
# To prevent an actual crash, I will wrap this in a try-except block.
try:
    print(f"Square root of 16: {calculate_sqrt_buggy()}")
except ImportError as e:
    print(f"Caught an error: {e}")
except AttributeError as e:
    print(f"Caught an AttributeError (if 'maths' was somehow imported but has no 'sqrt'): {e}")


Buggy code output (will eventually raise ImportError):
Caught an error: No module named 'maths'


As you can see, attempting to `import maths` results in an `ImportError` because there is no module named 'maths' in Python's standard library or installed packages. The correct module name is `math`.

### The Fix

To fix this `ImportError`, simply correct the typo in the import statement from `maths` to `math`. This will allow Python to correctly locate and import the standard math module, making its functions (like `sqrt`) available for use.

In [26]:
# Corrected code
def calculate_sqrt_fixed():
    import math # Correct module name
    return math.sqrt(16)

print("\nCorrected code output:")
print(f"Square root of 16: {calculate_sqrt_fixed()}")



Corrected code output:
Square root of 16: 4.0


#observation

(Import Error – Wrong Module Usage) has been successfully addressed! We illustrated how a common typo in an import statement can lead to an ImportError and demonstrated the fix by using the correct module name (math instead of maths).